ir-compiler.ts

```diff
@@ -0,0 +1,256 @@
+    newOffset += 1;
+  });
+  prog.classes.forEach(cls => {
+    const classFields = new Map();
+    cls.fields.forEach((field, i) => classFields.set(field.name, [i, field.value]));
+    newClasses.set(cls.name, classFields);
+  });
+  return {
+    globals: newGlobals,
+    classes: newClasses,
+    locals: env.locals,
+    offset: newOffset
+  }
+}
+
+export function makeLocals(locals: Set<string>) : Array<string> {
+  const localDefines : Array<string> = [];
+  locals.forEach(v => {
+    localDefines.push(`(local $${v} i32)`);
+  });
+  return localDefines;
+}
+
+export function compile(ast: Program<Type>, env: GlobalEnv) : CompileResult {
+  const withDefines = augmentEnv(env, ast);
+
+  const definedVars : Set<string> = new Set(); //getLocals(ast);
+  definedVars.add("$last");
+  definedVars.forEach(env.locals.add, env.locals);
+  const localDefines = makeLocals(definedVars);
+  const funs : Array<string> = [];
+  ast.funs.forEach(f => {
+    funs.push(codeGenDef(f, withDefines).join("\n"));
+  });
+  const classes : Array<string> = ast.classes.map(cls => codeGenClass(cls, withDefines)).flat();
+  const allFuns = funs.concat(classes).join("\n\n");
+  // const stmts = ast.filter((stmt) => stmt.tag !== "fun");
+  const inits = ast.inits.map(init => codeGenInit(init, withDefines)).flat();
+  const commandGroups = ast.stmts.map((stmt) => codeGenStmt(stmt, withDefines));
+  const commands = localDefines.concat(inits.concat([].concat.apply([], commandGroups)));
+  withDefines.locals.clear();
+  return {
+    functions: allFuns,
+    mainSource: commands.join("\n"),
+    newEnv: withDefines
+  };
+}
+
+function codeGenStmt(stmt: Stmt<Type>, env: GlobalEnv): Array<string> {
+  switch (stmt.tag) {
+    case "assign":
+      var valStmts = codeGenExpr(stmt.value, env);
+      if (env.locals.has(stmt.name)) {
+        return valStmts.concat([`(local.set $${stmt.name})`]); 
+      } else {
+        const locationToStore = [`(i32.const ${envLookup(env, stmt.name)}) ;; ${stmt.name}`];
+        return locationToStore.concat(valStmts).concat([`(i32.store)`]);
+      }
+
+    case "return":
+      var valStmts = codeGenValue(stmt.value, env);
+      valStmts.push("return");
+      return valStmts;
+
+    case "expr":
+      var exprStmts = codeGenExpr(stmt.expr, env);
+      return exprStmts.concat([`(local.set $$last)`]);
+
+    case "pass":
+      return []
+
+    case "field-assign":
+      var objStmts = codeGenValue(stmt.obj, env);
+      var objTyp = stmt.obj.a;
+      if(objTyp.tag !== "class") { // I don't think this error can happen
+        throw new Error("Report this as a bug to the compiler developer, this shouldn't happen " + objTyp.tag);
+      }
+      var className = objTyp.name;
+      var [offset, _] = env.classes.get(className).get(stmt.field);
+      var valStmts = codeGenValue(stmt.value, env);
+      return [
+        ...objStmts,
+        `(i32.add (i32.const ${offset * 4}))`,
+        ...valStmts,
+        `(i32.store)`
+      ];
+
```

Comment:
Is the plan for the cases below to do more or less a while/if with an integer to simulate basic blocks? Or do something clever like Stackifier?

https://medium.com/leaningtech/solving-the-structured-control-flow-problem-once-and-for-all-5123117b1ee2

Reply:
One option is to do the switch/loop and have it be a project to improve it!